### PR TITLE
Add support for more `CausalLM`s

### PR DESCRIPTION
**Pull request by jon-tow**
This PR adds support for more `CausalLM`s in the HuggingFace Hub. Previously, only models that followed the `gpt2` architecture layer-naming convention were supported (except for ILQL which also supports `gpt-neox`).
This will allow one to use models such as OPT and Pythia out of the box.

`wandb` reports:
* [ILQL sentiment](https://wandb.ai/jon-tow/trlx/reports/trlx-ILQL-Sentiment-Benchmark--VmlldzozMDkwNDM0?accessToken=taz92kkckt9meyjyqvjs6deztadvake6ugzfk31nocrd6hpcx0po586b3fmx1dpb)
* [PPO sentiment](https://wandb.ai/jon-tow/trlx/reports/trlx-PPO-Sentiment-Benchmark--VmlldzozMDkwNDIx?accessToken=1xdjyn26hw44gb6m9lel9c87m1dwptdifewrzzhw2vo98n8dof9keu75iulhmh53)

Notes:
- As discussed in the TRLX meeting (Dec. 5, 2022) - I've ignored the implementation of BLOOM for hydra-style model branching. CC @LouisCastricato 

Related Issue: #121

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import transformers
+
+import trlx.utils.modeling as modeling_utils
+
+
+# Test modeling utils
+
+
+@pytest.mark.parametrize(
+    "model_name",
```
Comment (Dahoas) [inline tests/test_utils.py:12]: Nice this is great

File: docs/source/models.rst [hunk 2]
```diff
@@ -19,13 +19,13 @@ Note that new models must be registered with ``trlx.model.register_model``.
 .. autoclass:: trlx.model.accelerate_ppo_model.AcceleratePPOModel
```
Comment (Dahoas) [inline docs/source/models.rst:19]: You will probably have to merge with @reciprocated latest commit (just fyi)

File: trlx/utils/modeling.py [hunk 3]
```diff
@@ -1,11 +1,129 @@
...
+
+def rhasattr(obj, attr):
+    """A chain-able attribute version of hasattr. For example, to check if
+    `obj` has the attribute `foo.bar.baz`, you can use:
+        `rhasattr(obj, "foo.bar.baz")`
+    Reference: https://stackoverflow.com/a/31174427
+    """
+    _nested_attrs = attr.split(".")
+    _curr_obj = obj
+    for _a in _nested_attrs[:-1]:
```
Comment (Dahoas) [inline trlx/utils/modeling.py:32]: Why [:-1]?

**Comment by Dahoas** [comment]
Looks very nice! For some reason tests for the HydraHead are breaking

File: trlx/utils/modeling.py (hunk 3 above)
Comment (jon-tow) [reply trlx/utils/modeling.py:32]: For some weird reason, nesting `hasttr`s doesn't always identify the attributes correctly. This walks through the dot-accesses to find the next to last object attribute that can be used to grab the proper boolean `hasattr` value.

**Comment by jon-tow** [comment]
@Dahoas Thanks for the review! This is still in draft as I'll be adding more tests and running examples. I'll bother you again for another review later, if you don't mind 😄

**Comment by dongs0104** [comment]
OPTForCausalLM is working well ? cause i changed ppo trainer code, and trained but result was bad results was not like gpt2,

OPT Model  is going to add a special token, `</s>` value function loss is not convergence. any idea for it ?

**Comment by jon-tow** [comment]
Hi @dongs0104! The default PPO trainer uses a GPT [hydra-head](https://github.com/CarperAI/trlx/blob/871cfba02b7c4f94ec7c929f875d0e044ac79db7/trlx/model/nn/ppo_models.py#L496) modification which will require re-working of the internals to support other models such as OPT. I plan to add hydras for OPT in this PR soon.

Thanks for bringing awareness to possible special-token issues - we'll keep an eye on it 👍

**Comment by danyang-rainbow** [comment]
Good job!

File: trlx/model/nn/ilql_models.py [hunk 4]
```diff
@@ -265,8 +264,8 @@ def generate(
         temperature=1,
         top_k=20,
         logit_mask=None,
-        pad_token_id=50256,
-        eos_token_id=50256,
+        pad_token_id=None,
```
Comment (LouisCastricato) [inline trlx/model/nn/ilql_models.py:267]: Why the change

**Review by LouisCastricato — APPROVED** [review]
Looks great! Lets merge.

**Comment by Dahoas** [comment]
This is excellent!

File: trlx/model/nn/ilql_models.py (hunk 4 above)
Comment (jon-tow) [reply trlx/model/nn/ilql_models.py:267]: `50256` is the default `eos_token_id` for `gpt2` based tokenizers - this just ensures there is no implicit default for the other non-gpt2 models.
